java-frontend/src/main/java/org/sonar/java/regex/JavaCharacterParser.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.regex;
+
+import org.sonar.java.regex.ast.IndexRange;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.RegexSource;
+
+/**
+ * Parse the contents of string literals and provide the individual characters of the string after processing escape
+ * sequences
+ */
+public class JavaCharacterParser {
+
+  public static final int EOF = -1;
+
+  private final RegexSource source;
+
+  private final JavaUnicodeEscapeParser unicodeProcessedCharacters;
+
+  private JavaCharacter current;
+
+  public JavaCharacterParser(RegexSource source) {
+    this.source = source;
+    this.unicodeProcessedCharacters = new JavaUnicodeEscapeParser(source);
+    moveNext();
+  }
+
+  public void moveNext() {
+    current = parseJavaCharacter();
+  }
+
+  public JavaCharacter getCurrent() {
+    return current;
+  }
+
+  public int getCurrentChar() {
+    if (current != null) {
+      return current.getCharacter();
+    } else {
+      return EOF;
+    }
+  }
+
+  public IndexRange getCurrentIndexRange() {
+    if (current != null) {
+      return current.getRange();
+    } else {
+      return new IndexRange(source.length(), source.length());
+    }
+  }
+
+  public int getCurrentStartIndex() {
+    if (current == null) {
+      return source.length();
+    } else {
+      return current.getRange().getBeginningOffset();
+    }
+  }
+
+  public boolean isAtEnd() {
+    return current == null;
+  }
+
+  public boolean isNotAtEnd() {
+    return current != null;
+  }
+
+  public boolean currentIs(char ch) {
+    return current != null && current.getCharacter() == ch;
+  }
+
+  private JavaCharacter parseJavaCharacter() {
+    JavaCharacter javaCharacter = unicodeProcessedCharacters.getCurrent();
+    if (javaCharacter == null) {
+      return null;
+    }
+    if (javaCharacter.getCharacter() == '\\') {
+      return parseJavaEscapeSequence(javaCharacter);
+    }
+    unicodeProcessedCharacters.moveNext();
+    return javaCharacter;
+  }
+
+  private JavaCharacter parseJavaEscapeSequence(JavaCharacter backslash) {
```

Comment:
Should be annotated with `@CheckForNull` as well, when encountering invalid string literals.